### PR TITLE
Add Seq# based optimistic concurrency control to UpdateRequest

### PR DESCRIPTION
**Pull request by bleskes**
The update request has a lesser known support for a one off update of a known document version. This PR adds an a seq# based alternative to power these operations.

Relates #36148 
Relates #10708

### CODE REVIEW

**Review by ywelsch — APPROVED** [review]
Thanks for picking this up @bleskes

**Review by dnhatn — APPROVED** [review]
I left some minor nits. LGTM.

File: docs/reference/docs/update.asciidoc [hunk 1]
```diff
@@ -349,3 +349,11 @@ version numbers being out of sync with the external system.  Use the
 <<docs-index_,`index` API>> instead.
 
 =====================================================
+
+`if_seq_no` and `if_primary_term`::
+
+Update operations can be made optional and only be performed if the last
```
[the remark this answers is not in the document]
Comment (bleskes) [reply docs/reference/docs/update.asciidoc]: good question. I took this over from the text in IndexRequest. I can change all if you prefer.

File: server/src/main/java/org/elasticsearch/action/update/UpdateRequest.java [hunk 2]
```diff
@@ -170,6 +179,27 @@ public ActionRequestValidationException validate() {
...
 
+        if (ifSeqNo != UNASSIGNED_SEQ_NO && (
+            versionType != VersionType.INTERNAL || version != Versions.MATCH_ANY
+        )) {
+            validationException = addValidationError("compare and write operations can not use versioning", validationException);
+        }
+        if (ifPrimaryTerm == UNASSIGNED_PRIMARY_TERM && ifSeqNo != UNASSIGNED_SEQ_NO) {
+            validationException = addValidationError("ifSeqNo is set, but primary term is [0]", validationException);
+        }
+        if (ifPrimaryTerm != UNASSIGNED_PRIMARY_TERM && ifSeqNo == UNASSIGNED_SEQ_NO) {
```
[the remark this answers is not in the document]
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/action/update/UpdateRequest.java]: yeah. I now changed to take another approach which I also took in a future PR for something else. Let me know if you prefer it.

File: server/src/main/java/org/elasticsearch/action/update/UpdateRequest.java [hunk 3]
```diff
@@ -531,6 +561,55 @@ public VersionType versionType() {
         return this.versionType;
     }
 
+    /**
+     * only perform this update request if the document was last modification was assigned the given
```
[the remark this answers is not in the document]
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/action/update/UpdateRequest.java]: fixed. thanks

File: server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 4]
```diff
@@ -1378,6 +1378,38 @@ public void testVersionedUpdate() throws IOException {
...
+        Engine.Index create = new Engine.Index(newUid(doc), primaryTerm.get(), doc, Versions.MATCH_DELETED);
+        Engine.IndexResult indexResult = engine.index(create);
+        if (randomBoolean()) {
+            engine.refresh("test");
+        }
+        if (randomBoolean()) {
+            engine.flush();
+        }
+        try (Engine.GetResult get = engine.get(
+            new Engine.Get(true, false, doc.type(), doc.id(), create.uid())
```
[the remark this answers is not in the document]
Comment (bleskes) [reply server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: good catch. We should get from the translog all the time. I randomized the flush/refeshes before

File: server/src/main/java/org/elasticsearch/index/get/ShardGetService.java [hunk 5]
```diff
@@ -113,7 +117,7 @@ public GetResult getForUpdate(String type, String id, long version, VersionType
     public GetResult get(Engine.GetResult engineGetResult, String id, String type,
                             String[] fields, FetchSourceContext fetchSourceContext) {
         if (!engineGetResult.exists()) {
-            return new GetResult(shardId.getIndexName(), type, id, SequenceNumbers.UNASSIGNED_SEQ_NO, 0, -1, false, null, null);
+            return new GetResult(shardId.getIndexName(), type, id, UNASSIGNED_SEQ_NO, 0, -1, false, null, null);
```
[the remark this answers is not in the document]
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/get/ShardGetService.java]: +1

File: server/src/main/java/org/elasticsearch/index/get/ShardGetService.java [hunk 6]
```diff
@@ -162,14 +166,14 @@ private GetResult innerGet(String type, String id, String[] gFields, boolean rea
...
-                    .version(version).versionType(versionType));
+                    .version(version).versionType(versionType).setIfSeqNo(ifSeqNo).setIfPrimaryTerm(ifPrimaryTerm));
             if (get.exists() == false) {
                 get.close();
             }
         }
 
         if (get == null || get.exists() == false) {
-            return new GetResult(shardId.getIndexName(), type, id, SequenceNumbers.UNASSIGNED_SEQ_NO, 0, -1, false, null, null);
+            return new GetResult(shardId.getIndexName(), type, id, UNASSIGNED_SEQ_NO, 0, -1, false, null, null);
```
[the remark this answers is not in the document]
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/get/ShardGetService.java]: +1

File: server/src/main/java/org/elasticsearch/index/engine/Engine.java [hunk 7]
```diff
@@ -1602,6 +1612,26 @@ public Get versionType(VersionType versionType) {
         public boolean isReadFromTranslog() {
             return readFromTranslog;
         }
+
+
+        public Get setIfSeqNo(long seqNo) {
```
[the remark this answers is not in the document]
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/engine/Engine.java]: I thought people wanted to migrate towards get/set. there is already a mixture in this file.

File: docs/reference/docs/update.asciidoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (ywelsch) [reply docs/reference/docs/update.asciidoc]: ++ to change it all

File: server/src/main/java/org/elasticsearch/action/update/UpdateRequest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/action/update/UpdateRequest.java]: looks good

**Comment by bleskes** [comment]
Thx @ywelsch and @dnhatn

**Pull request merged**